File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/CAEmitterCell粒子动画/大雪效果/WGSnowAniViewController.m
//
//  WGSnowAniViewController.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/23.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGSnowAniViewController.h"

@interface WGSnowAniViewController ()

@property (nonatomic, strong) CAEmitterLayer *emitterLayer;
@property (nonatomic, strong) CAEmitterCell *emitterCell;

@end

@implementation WGSnowAniViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"大雪效果";
    self.view.backgroundColor = UIColor.blackColor;
    
    //粒子发射器图层
    self.emitterLayer = [CAEmitterLayer layer];
    //粒子发射器位置
    _emitterLayer.emitterPosition = CGPointMake(0, -30);
    //粒子发射器的范围
    _emitterLayer.emitterSize = CGSizeMake(SCREEN_WIDTH*2, 0);
    //发射模式
    _emitterLayer.emitterMode = kCAEmitterLayerOutline;
    //粒子模式
    _emitterLayer.emitterShape = kCAEmitterLayerLine;
    
    self.emitterCell = [CAEmitterCell emitterCell];
    _emitterCell.contents = (__bridge id _Nullable)([UIImage imageNamed:@"snow"].CGImage);
    //设置粒子缩放比例
    _emitterCell.scale = 0.02;
    //缩放范围
    _emitterCell.scaleRange = 0.5;
    //每秒粒子产生数量
    _emitterCell.birthRate = 100;
    //粒子生命周期
    _emitterCell.lifetime = 80;
    //粒子透明速度
    _emitterCell.alphaSpeed = -0.01;
    //粒子速度
    _emitterCell.velocity = 80;
    _emitterCell.velocityRange = 50;
    //设置角度
    _emitterCell.emissionRange = M_PI;
    //设置粒子旋转角速度
    _emitterCell.spin = M_PI_4;
    
    _emitterLayer.emitterCells = @[_emitterCell];
    
    [self.view.layer addSublayer:_emitterLayer];
}



@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/粒子动画/大雪效果/WGSnowAniViewController.swift
//
//  WGSnowAniViewController.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/5/6.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGSnowAniViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = UIColor.black
        title = "大雪动画"
        
        //粒子发射器图层
        let emitterLayer = CAEmitterLayer()
        //粒子发射器位置
        emitterLayer.emitterPosition = CGPoint(x: 0, y: -30)
        //粒子发射器的范围
        emitterLayer.emitterSize = CGSize(width: WgWith*2, height: 0)
        //发射模式
        emitterLayer.emitterMode = .outline
        //粒子模式
        emitterLayer.emitterShape = .line
        
        let emitterCell = CAEmitterCell()
        emitterCell.contents = UIImage(named: "snow")?.cgImage
        //设置粒子缩放比例
        emitterCell.scale = 0.02;
        //缩放范围
        emitterCell.scaleRange = 0.5;
        //每秒粒子产生数量
        emitterCell.birthRate = 100;
        //粒子生命周期
        emitterCell.lifetime = 80;
        //粒子透明速度
        emitterCell.alphaSpeed = -0.01;
        //粒子速度
        emitterCell.velocity=80;
        emitterCell.velocityRange=50;
        //设置角度
        emitterCell.emissionRange = CGFloat(Double.pi);
        //设置粒子旋转角速度
        emitterCell.spin = CGFloat(Double.pi/4);
        
        emitterLayer.emitterCells = [emitterCell]
        
        view.layer.addSublayer(emitterLayer)
        
    }

}
